crates/accelerate/src/synthesis/linear_phase/cnot_phase_synth.rs

```diff
@@ -0,0 +1,428 @@
+                    self._i.clone().into_iter().filter(|&x| x != _j).collect(),
+                    _j,
+                ));
+            } else {
+                self.q.push((
+                    cnots1,
+                    self._i.clone().into_iter().filter(|&x| x != _j).collect(),
+                    self._ep,
+                ));
+            }
+            self.q.push((
+                cnots0,
+                self._i.clone().into_iter().filter(|&x| x != _j).collect(),
+                self._ep,
+            ));
+
+            self.next()
+        } else {
+            None
+        }
+    }
+}
+
+struct BindingIterator {
+    num_qubits: usize,
+    q: Vec<(Array2<u8>, Vec<usize>, usize)>,
+    phase_iter_handle: PhaseIterator,
+    cx_phase_iter_handle: Option<CXPhaseIterator>,
+    phase_iterator_done: bool,
+}
+
+impl BindingIterator {
+    fn new(s: Array2<u8>, angles: Vec<String>, state: Array2<u8>) -> Self {
+        let num_qubits = s.nrows();
+        let q = vec![(
+            s.clone(),
+            (0..num_qubits).collect::<Vec<usize>>(),
+            num_qubits,
+        )];
+        Self {
+            num_qubits,
+            q,
+            phase_iter_handle: PhaseIterator::new(num_qubits, s, state, angles),
+            cx_phase_iter_handle: None,
+            phase_iterator_done: false,
+        }
+    }
+}
+
+impl Iterator for BindingIterator {
+    type Item = Instruction;
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.phase_iterator_done {
+            let data = self.phase_iter_handle.next();
+            if data.is_none() {
+                self.cx_phase_iter_handle = Some(CXPhaseIterator::new(
+                    self.num_qubits,
+                    self.phase_iter_handle.s_cpy.clone(),
+                    self.phase_iter_handle.state.clone(),
+                    self.phase_iter_handle.rust_angles.clone(),
+                    self.q.clone(),
+                ));
+                self.phase_iterator_done = true;
+                self.next()
+            } else {
+                data
+            }
+        } else if let Some(handle) = self.cx_phase_iter_handle.as_mut() {
+            handle.next()
+        } else {
+            None
+        }
+    }
+}
+
+/// This function implements a Gray-code inspired algorithm of synthesizing a circuit
+/// over CNOT and phase-gates with minimal-CNOT for a given phase-polynomial.
+/// The algorithm is described as "Gray-Synth" algorithm in Algorithm-1, page 12
+/// of paper "https://arxiv.org/abs/1712.01859".
+#[pyfunction]
+#[pyo3(signature = (cnots, angles, section_size=2))]
+pub fn synth_cnot_phase_aam(
+    py: Python,
+    cnots: PyReadonlyArray2<u8>,
+    angles: &Bound<PyList>,
+    section_size: Option<i64>,
+) -> PyResult<CircuitData> {
+    let s = cnots.as_array().to_owned();
+    let num_qubits = s.nrows();
+
+    let rust_angles = angles
+        .iter()
+        .filter_map(|data| data.extract::<String>().ok())
```

Comment:
doing 
```rust
.filter_map(|data| data.extract::<String>().or_else(|_| data.extract::<f64>().map(|f| f.to_string())).ok())
```
or, doing 
```python
angles = [angle if isinstance(angle, str) else f"{angle}" for angle in angles]
```
in `qiskit/synthesis/linear_phase/cnot_phase_synth.py`

takes the same amount of time to execute, so, I decided to keep this check in python space only! :)